src/core/geometry/qgslinestring.cpp

```diff
@@ -587,6 +589,31 @@ QgsRectangle QgsLineString::calculateBoundingBox() const
   return QgsRectangle( xmin, ymin, xmax, ymax, false );
 }
 
+QgsBox3d QgsLineString::calculateBoundingBox3d() const
+{
+  if ( mX.empty() )
+    return QgsBox3d();
+
+  auto result = std::minmax_element( mX.begin(), mX.end() );
```

Comment:
This can be further optimised -- check first if mBoundingBox.isNull(). If not, take the min/max x/y directly from that. (mBoundingBox is the cached 2d bounding box)

Reply:
ok